### ISSUE TITLE
[UX] Allow bulk generation of path aliases for individual entity bundles

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#2077]
See https://github.com/backdrop/backdrop-issues/issues/940#issuecomment-240263939

---

Steps for testing:
* set default URL alias pattern for posts to posts/[node:title] at admin/config/search/path/patterns
* devel generate content 
* confirm post paths are posts/[node:title]
* bulk-delete all aliases at admin/config/search/path/bulk-update
* confirm post paths are node/NID
* change default URL alias pattern for posts to articles/[node:title] at admin/config/search/path/patterns
* bulk-generate aliases for posts at admin/config/search/path/bulk-update
* confirm post paths are articles/[node:title]
* change default URL alias pattern for posts to blog/[node:title] at admin/config/search/path/patterns
* bulk-update aliases for posts at admin/config/search/path/bulk-update
* confirm post paths are blog/[node:title]

---
PR from @docwilmot ~~https://github.com/backdrop/backdrop/pull/1489~~
Combined PR from @quicksketch: https://github.com/backdrop/backdrop/pull/1533

**Comment by docwilmot** [comment]
Initial PR only for content types, not Taxonomy or Users yet. This is an initial proof of concept.

**Comment by klonos** [comment]
I quickly checked the sandbox and liked what I saw. A few suggestions/ideas:

1. We add a drop down menu and have the actions on that form (create/update/delete) be like bulk actions. So, "With selected: create URL aliases / delete URL aliases".

2. "Delete existing" should be "Update existing" instead.

3. Wouldn't it be better if the parent checkboxes acted as select-al/unselect-all for the children checkboxes? In that case, we should add a global "All URL aliases" top-most checkbox and nest everything under it.

4. Get rid of the "Delete URL aliases" tab.

**Comment by klonos** [comment]
Here's mockups...

Bulk operations fieldset and global, top-most "All URL aliases" checkbox added. "Delete" tab and form submit button removed:

![backdrop-issue2077-mockup-bulk_operations_options](https://cloud.githubusercontent.com/assets/2423362/17835267/37c782ae-67a8-11e6-9a09-ccb3f0e9745f.png)

...bulk deletion selected as action:

![backdrop-issue2077-mockup-bulk_operations_option-delete](https://cloud.githubusercontent.com/assets/2423362/17835276/a27bcc36-67a8-11e6-888d-5785fdae5e91.png)

...bulk creation selected as action reveals the "Update existing" column with individual checkboxes per bundle:

![backdrop-issue2077-mockup-bulk_operations_option-create](https://cloud.githubusercontent.com/assets/2423362/17835282/e8b9a646-67a8-11e6-859a-55ac73a83ef8.png)

**Comment by docwilmot** [comment]
thanks
1. I think the presence of the "Update existing" checkbox might make having a delete dropdown option confusing. I like the separate tab for deleting myself, but we can wait for more opinions here.
2. Agreed
3. Also agreed, in fact thats already done.
4. Only if 1 is the consensus, of course

**Comment by klonos** [comment]
>1. I think the presence of the "Update existing" checkbox might make having a delete dropdown option confusing.

That's why in my second mockup (where the delete action is selected from the drop-down menu), the "Update existing" column is not present at all. It is only shown when one selects the create action.

> 3. Also agreed, in fact thats already done.

Just gave the sandbox another go and there is room for improvement with the JS magic for selecting checkboxes:

a) Add an "All URL aliases" master checkbox that (de)selects *ALL* bundles (content + taxonomy + user)

b) Selecting all content types individually does not automatically select the parent "Content URL aliases" checkbox. I realize that it'd be silly for somebody to do this (especially when they have many content types), but shouldn't it be so that doing so makes the parent checkbox selected as well?

c) Related to point b above:
  - Check the parent "Content URL aliases" checkbox -> all content type checkboxes are checked too.
  - Uncheck the parent "Content URL aliases" checkbox -> all content type checkboxes are unchecked too.
  - Check the parent "Content URL aliases" checkbox again-> all content type checkboxes are checked too.
  - Uncheck one of the content types -> the parent "Content URL aliases" checkbox is unchecked (feels right)
  - Check the content type that you unchecked in the previous step -> the parent "Content URL aliases" checkbox remains unchecked (doesn't feel right)

**Comment by klonos** [comment]
I believe what we need for b and c is the "indeterminate" state of the HTML checkboxes:

https://css-tricks.com/indeterminate-checkboxes
https://stackoverflow.com/questions/1726096/tri-state-check-box-in-html
https://vanderlee.github.io/tristate

**Comment by klonos** [comment]
...updated issue summary with link to respective d.org issue.

**Comment by docwilmot** [comment]
The indeterminate plugin is cool, but not necessary for this application I think. That could be a feature request, and probably a contrib one at that IMHO.
And the Drupal issue also is a different issue to do with the `Update action` on the `admin/config/search/path/patterns/settings` page. Nothing to do with the bulk generation page.
I've updated the code for  b and c above.
Ideally, if we're going to have a "Select all" checkbox, we should just be using the `tableselect` form element instead of `theme_table`. I had tried that first and couldnt figure out how to combine that with having form elements within the table (the select boxes). Will try again later.

**Comment by docwilmot** [comment]
:smile: 

Interesting complication:
Seems theres a bit of a 'bug' with tableselect. When you check the 'select all' checkbox, the js in `tableselect.js` searches for *any* checkbox in the table and selects all of them. In this case, it selects all the "Update existing" checkboxes too!

I think this is a bug because the code should probably be made more specific, to search only for the row selection checkboxes.

So options from here:
1. Use tableselect and somehow try and override its js behaviour. I dont think I can manage that
2. Steal the js code from `tableselect.js` that puts the select all checkbox in the header.
3. Do a separate select all checkbox as per @klonos mockup above
4. Don't bother with a select all checkbox right now.

@quicksketch @jenlampton @serundeputy etc (does GitHub have a @everbodyinthisgroup option?) what would you do here?

**Comment by jenlampton** [comment]
I vote for #4, first get it done, then make it better :)

**Comment by jenlampton** [comment]
Also, I *love* what I'm seeing in this issue. I think this will be a huge UX win for managing URL aliases :)  

> We add a drop down menu and have the actions on that form (create/update/delete) be like bulk actions. So, "With selected: create URL aliases / delete URL aliases".

yes! 👍 

> "Delete existing" should be "Update existing" instead.

Why not all three? I've had use cases for create, update, and delete.

> Get rid of the "Delete URL aliases" tab.

Love it :) 

> Don't bother with a select all checkbox right now.

We might be able to fake this for now by checking/unchecking the child boxes when the parent box is checked using #states on the form. Have you tried that instead of using tableselect.js?

**Comment by docwilmot** [comment]
- added bulk options select
- added ability to delete aliases using checkboxes on same form
- hide update checkboxes when generating
- can now delete aliases for node/taxonomy subtypes (page, post), rather thn just all "content" 

last option required adding a new key to `hook_path_info()` and a delete callback. It is difficult to do deleteion of aliases for any arbitrary core/contrib entity without knowing how it stores its path strings in the `url_alias` table. the callback allows individual entities to tell Backdrop how to delete aliases for this entity.

**Comment by jenlampton** [comment]
Brilliant.

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by docwilmot** [comment]
Tagged 1.5.0 for review.

**Comment by klonos** [comment]
Looks amazing!! Good job @docwilmot 👍 

...but as always, I did some tire-kicking and found glitches:

1. leave the default action to "generate"
2. select the "all alias" checkbox->
  - all aliases checkboxes are selected [OK]
  - all "update existing" checkboxes are enabled [OK]
3. deselect either *one* of the two available options for content aliases ->
  - the parent "content type URL aliases" checkbox is deselected [OK]
  - the respective "update existing" checkboxes for the unselected content type and its parent checkbox are disabled [OK]
  - the top-level checkboxes for "taxonomy term aliases" and the "user account aliases" remain enabled [OK]
  - *BUT* their "update existing" checkboxes are disabled [BUG]
4. try re-selecting the content type alias checkbox you deselected in step 3 above ->
  - the top-level parent checkbox for content types is also selected [OK]
  - the respective "update existing" checkboxes are all enabled again [OK]
5. deselect the content type alias checkbox once again to reproduce the bug from step 3 above
6. this time select the top-level alias checkbox for content types ->
  - all content type alias checkboxes are selected [OK]
  - *BUT* the "update existing" checkboxes for the rest of the top-level aliases groups (terms and users) are disabled [BUG]

**Comment by klonos** [comment]
...to test the above bugs better, I've created a second vocabulary in the PR sandbox. The same bug occurs if you deselect any *one* of the vocabularies: all the "update existing" checkboxes for the top-level alias groups (content types, user accounts) get disabled.

**Comment by klonos** [comment]
Another one:

1. select all aliases in generation mode
2. select some of the "update existing" checkboxes
3. refresh the page ->
  - checked "update existing" boxes are remembered [OK]
  - all "update existing" boxes (both checked and unchecked) are in disabled state [BUG]

**Comment by klonos** [comment]
...also can I please haz a "select all" checkbox for the update existing column? Thanx 😉

**Comment by klonos** [comment]
Minor consistency nit... in both `/admin/people` and `/admin/content/node` we have an "Operations" fieldset legend and no label for the drop-down actions menu. Do the same here?

**Comment by docwilmot** [comment]
Sorry to be pushy, but I think this is small and doable for 1.5, I just need validation from @quicksketch @jenlampton re the code approach (I am not a coder!) before I can finalise and fix tests. And we are running out of time.

Note that this is "blocking" #940.

**Comment by jenlampton** [comment]
I looked over the PR and it looks pretty good to me. I wonder if we should batch delete the aliases like we do a batch for updating. There could be a lot of things getting deleted and we don't want to crash sites, but the query executed looks pretty simple.  I'm not sure when to batch or not to batch!  Let's see what @quicksketch thinks. 

> I am not a coder!

@docwilmot you are going to have to stop saying this. You are actually quite a good coder, like it or not! :) Also you are full of lots of good ideas...  so @docwilmot++ :)

**Comment by docwilmot** [comment]
> You are actually quite a good coder, like it or not! :) Also you are full of lots of good ideas

I'm framing a screenshot of this. Thanks @jenlampton.

As of last commit, tests are all green now, but currently woefully inadequate. It seems we never even had any tests for the UI bulk delete facility!! Would file a new issue, but I think i can just create those new tests to fix that with this PR. 

I've removed the JQuery magic to make the reset checkboxes disabled when the alias checkboxes are unchecked. Would be nice but its driving me crazy. Hopefully can be done later?

> I wonder if we should batch delete the aliases like we do a batch for updating.

The current Backdrop/D7 doesnt batch deletes either. Its a straight `DELETE from url_alias` query though, and no hooks firing. I dont know how many aliases you'd need to have to crash it. As you say on to @quicksketch.

**Comment by klonos** [comment]
> I am not a coder!

If that implies something like 0 coding skills, then I'm in the minus scale. What is the negative for coder? 😜

**Comment by jenlampton** [comment]
This whole sub-thread makes me happy :) With you two 'non-coders' becoming coders, that means that Backdrop is becoming successful! It's appealing / inviting to people who consider themselves non-coders! WIN.

**Comment by quicksketch** [comment]
This PR is looking great!

I tried it out and everything worked just as I would expect. Even redirects were created when making the new aliases that replaced old ones. Very nice!

This was about the only inaccuracy I found:

> No new URL aliases generated; 3 aliases were updated.

But there was only a single piece of content (I updated aliases for all content types). So only one alias was updated. I also wouldn't call it "updated" if a redirect is being created. I'd say "1 new alias created. 1 redirect created." But I understand that the effect of "updating" an alias has a different effect based on what setting you have specified in the URL pattern settings, and getting those counts could be tricky. But at the very least, it seems like either more aliases are getting updated than they should or the message is somehow inaccurate.

A bigger issue that came up in the code itself: A large number of nodes (or other type of content) would cause such a long query (imagine a query with an `IN()` statement that had 10,000 rows) it may cause an issue. I think batching would be safer.

**Comment by quicksketch** [comment]
Actually after some Googling, I think we may be fine with the `IN()` statement after all. 

http://stackoverflow.com/a/8808935/845793

`The number of values in the IN list is only limited by the max_allowed_packet value.`

The max packet variable is usually pretty big > 16MB is typical. Drupal/Backdrop sites need a big value in the first place to store things like FomAPI arrays. With 4 bytes per character and each node system path `node/[nid]` taking about 16 characters, that'd be: 16777216 (16 MB) / (12 * 4) = 349525.

So you could probably update about 350,000 nodes at once before you ran into this problem.

I don't know... you tell me. Should we accept that as-is and then make a follow-up for batching?

**Comment by jenlampton** [comment]
> Should we accept that as-is and then make a follow-up for batching?

+1 for baby steps.

**Comment by quicksketch** [comment]
Overall the new code looks great. I personally find the UI for updating checkboxes still a little overwhelming, but it's better than the previous attempts. If this is going to unblock setting default patterns then I'm up for it. I had to do a small amount of cleanup, so I made a new PR made at https://github.com/backdrop/backdrop/pull/1533.

Changes:
- New `path_generate.js` moved to `path.admin.js`.
- New `path.css` moved to `path.admin.css`.
- Fixed jQuery variables in the global namespace and added $.once() checking to prevent double-binding of events.
- Added RTL styles and tidied up the CSS to be page-specific in the event `path.admin.css` gets used elsewhere at some point.

**Comment by jenlampton** [comment]
I'd be down for giving this a review as soon as I can get my tests passing for https://github.com/backdrop/backdrop-issues/issues/1152 :)

**Comment by jenlampton** [comment]
This is fantastic!!! I love the new feature and can't wait to use it on my own sites :) Thank you @docwilmot for putting in all the work to get it ready, and @quicksketch for cleaning up the JQuery magic :)

When I started testing the PR I initially thought the ability to delete had been removed. It hasn't,  but people are likely to think think it has because:
* The 4th/last "delete" tab is now missing
* The other tab "Update URL aliases" didn't register as having changed from "Bulk generate"
* When scanning the update page, the word "Delete" is not visible anywhere on the page. This is because "Delete" is the second option in the Bulk Actions drop-down, so hidden until dropped-open.
* Additionally, once you realize that you can delete on the update page, the heading for the type-of-aliases column remains `Generate`, making you think twice before trying to delete - is that going to generate when I want to delete?

##Recommended improvements and other small nits:
* Change the tab name to something like `URL alias bulk actions` - this will indicate delete as well as update
* Move the tab to the 2nd position, following List and preceding Patterns.  This puts the settings "at the end" similar to other UIs (layouts, views) and the change in tab order will also signal to people that this change is not simply the omission of the last `Delete URL aliases` tab.
* Heading for the `Bulk Actions` fieldset should be changed to `Operations` to be consistent with node, user, and other bulk-actions UI.
* Default value for select list should be `- Choose an action -` also for consistency
* The word "aliases" appears in both options the drop-down without the leading "URL" 
* ` URL alias pattern for all Page URLs ` should be changed to ` All Page URL aliases` (and maybe we make the content type name bold?)  - We don't want the patterns to be changed, but the aliases themselves.
* Column heading should be changed from `Generate` to `All URL aliases`. Especially with the checkbox next to it, we need to be clear what checking that box means.
* Confirm form text `Deleting these URL aliases cannot be undone.` should probably remain `This action cannot be undone` for the sake of translations. If we want to make these confirm-form messages more useful (and we should!) let's open a new issue and do it consistently throughout core. :)

I am on the fence about `Update existing`. It seems to me that it would be a better user experience if it were it's own Bulk operation, rather than a checkbox addition to 'Generate'.  What if I want to bulk update existing aliases without generating new ones?

But the again, I do like the ability to generate and update at the same time too,. Should the checkbox remain, or would it also be more natural to add a `Generate and update` bulk operation instead?

**Comment by quicksketch** [comment]
I also ran into most of the problems @jenlampton described above, but it seemed like this was the best UI we had come up with. Agreed though, even though this adds great new functionality, it has some pretty significant issues in usability. As a site user I would be a little perplexed with the double-checkboxes.

> I am on the fence about Update existing. It seems to me that it would be a better user experience if it were it's own Bulk operation, rather than a checkbox addition to 'Generate'. What if I want to bulk update existing aliases without generating new ones?

Although requiring more steps if you were bulk changing everything on your site, I like this idea to simply separate the concepts of generating vs. updating. I find this to be much more clear as separate operations:

- Bulk generate: Creates new aliases *only* on non-aliased content.
- Bulk update: Updates aliases *only* on already-aliased content.
- Bulk delete: Deletes aliases on already-aliased content.

Sure if I were doing this on my own site, I would probably execute the "Update" operation first, then "Generate" second. Although in truth you could do it in either order and you should get the same result.

**Comment by docwilmot** [comment]
Youre right about the checkboxes. And the separate bulk option makes sense. Is somebody on this or am I picking it up?

**Comment by quicksketch** [comment]
Awesome, looks like you rerolled the PR @docwilmot! Sorry for the delay, there's a lot going on in the issue queue right now. Marking needs review again.

**Comment by quicksketch** [comment]
@docwilmot nailed it in https://github.com/backdrop/backdrop/pull/1541. Great job! Merged into 1.x for 1.5.0. High five!
